draft-pda-core.md

```diff
@@ -291,11 +293,427 @@ Each PA task in this document is divided into three sub-protocols as follows.
+decrypt a share it receives.
+
+## Pre-conditions
+
+We assume the following conditions hold before the client begins uploading its
+data:
+1. The client, aggregators, and collector are configured with a specific PA task.
+1. The client knows the URL of the leader endpoint, e.g., `example.com/metrics`.
+   We write this URL as `[leader]` below. (We write `[helper]` for a helper's
+   URL.)
+1. The client and leader can establish a leader-authenticated secure channel.
+1. The leader and each helper can establish a leader-authenticated secure
+   channel.
+1. Each helper has chosen an HPKE key pair.
+1. The aggregators agree on a set of PA tasks, as well as the PA protocol and
+   parameters used for each task.
+
+[TODO: It would be clearer to include a "pre-conditions" section prior to each
+"phase" of the protocol.]
+
+## Upload {#pa-upload}
+
+[TODO: Add an illustration of this sub-protocol.]
+
+Uploading a report involves two requests to the leader. In the  *upload start
+request*, the client discovers the protocol-specific parameters it needs to
+generate the report, as well as the endpoint URL of each helper. In the *upload
+finish request*, it uploads its report to the leader.
+
+[NOTE: @acmiyaguchi pointed out that the use of an anonymizing proxy for
+uploading shares might be easier to implement if the "upload" phase involved a
+single HTTP request. However, OHTTP
+(https//www.ietf.org/archive/id/draft-thomson-http-oblivious-01.html) allows
+clients to make multiple requests through a proxy, so these kinds of use cases
+should work.]
+
+[NOTE: @cjpatton: Breaking the upload phase into two requests is useful for
+supporting Prio-like proof systems in which the leader sends the client a
+"challenge" it uses to generate the proof. (See the SIMD construction of
+{{BBCp19}}, Section 5.2. Here, the "challenge" is a randomly generated field
+element.) Because the client/leader are using TLS, another way to implement this
+would be to export a secret from the key schedule.]
+
+### Upload Start
+
+The client sends a POST request to `[leader]/upload_start` with the following
+message:
+
+~~~
+struct {
+  PATask task;
+} PAUploadStartReq;
+~~~
+
+The `task` field corresponds to the PA task for which a report will be
+generated.
+
+The leader responds to well-formed requests to `[leader]/upload_start` with
+status 200 and the following message:
+
+~~~
+struct {
+  PAParam param;
+  Url helper_urls<1..2^16-1>;
+  select (PAUploadStartResp.param.proto) {
+    case prio: PrioUploadStartResp;
+    case hits: HitsUploadStartResp;
+  }
+} PAUploadStartResp;
+
+opaque Url<1..2^16-1>;
+~~~
+
+The message includes the URL of each helper and any protocol-specific parameters
+the client needs to generate its report. The leader's response to malformed
+requests is specified in {{pa-error-common-aborts}}.
+
+### Upload Finish
+
+For each URL `[helper]` in `PAUploadStartResp.helper_urls`, the client sends a
```

Comment:
This seems like it has some obvious security problems. What stops the leader from just telling the client it runs all the helpers?

Reply:
Presumably clients will just check against their set of helpers or some local policy?

Reply:
But then why do I need the server to tell me.

I think there's a philosophical issue here: We should avoid having the leader tell the client anything that could be inconsistent with things it already needs to know. If we are trying to detect mismatch, we should instead have a config structure where the server sends the hash (or it's included in an AAD or something). That way we detect mismatch but don't rely on the client checking each parameter. Otherwise, we get into situations where the client might inappropriately trust the leader.

Reply:
That's a good point -- if the baseline assumption is that the leader is not trusted to not collude, and clients can't reasonably check for this, then this needs to change. (Let's do that in the next PR.)